zqd/api/connection.go

```diff
@@ -0,0 +1,364 @@
+// The body can be an io.Reader for the big stuff.  Body can also be an arbitrary
+// gc protocol object type-aligned to the expected parameter.
+func (c *Connection) _call(ctx context.Context, method, path string, hdr http.Header, body interface{}) (*http.Response, error) {
+	req, err := c.Request(method, path, hdr, body)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+	resp, err := c.Do(req)
+	if err != nil {
+		return resp, err
+	}
+	err = responseError(resp)
+	if err != nil {
+		return resp, err
+	}
+	return resp, nil
+}
+
+// call transmits a request to the server and waits for a response and
+// returns the response in the provided result object represented here
+// as an empty interface.  The request is constructed from method, path, body.
+// If a result is expected, then the result value is filled in.  Otherwise,
+// result should be nil.
+func (c *Connection) call(method, path string, hdr http.Header, body, result interface{}) (status int, err error) {
+	resp, err := c._call(context.Background(), method, path, hdr, body)
+	if resp != nil {
+		status = resp.StatusCode
+	}
+	if err != nil {
+		return status, err
+	}
+	if result == nil {
+		return status, nil
+	}
+	typ := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(typ, "application/json") {
+		return status, fmt.Errorf("expected response Content-Type header to be 'application/json', got %q", typ)
+	}
+	err = json.NewDecoder(resp.Body).Decode(result)
+	err2 := resp.Body.Close()
+	if err != nil {
+		return status, err
+	}
+	return status, err2
+}
+
+// stream transmits a request to the server and returns a bufio.Scanner that
+// parses double-newline chunks of data and returns them as byte slices.
+// It's up to the caller to wrap a decoder around this scanner.
+func (c *Connection) stream(ctx context.Context, method, path string, hdr http.Header, body interface{}) (*bufio.Scanner, error) {
+	resp, err := c._call(ctx, method, path, hdr, body)
+	if err != nil {
+		return nil, err
+	}
+	return NewJSONPipeScanner(resp.Body), nil
+}
+
+type ErrorResponse struct {
+	*http.Response
+	Err error
+}
+
+func responseError(resp *http.Response) error {
+	if resp.StatusCode < 400 {
+		return nil
+	}
+	resErr := &ErrorResponse{Response: resp}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if strings.HasPrefix(resp.Header.Get("Content-Type"), "application/json") {
+		var apierr *Error
+		err := json.Unmarshal(body, apierr)
+		if err != nil {
+			resErr.Err = err
+		} else {
+			resErr.Err = apierr
+		}
+	} else {
+		resErr.Err = errors.New(string(body))
+	}
+	return resErr
+}
+
+func (e *ErrorResponse) Unwrap() error {
+	return e.Err
+}
+
+func (e *ErrorResponse) Error() string {
+	return fmt.Sprintf("status code %d: %v", e.StatusCode, e.Err)
+}
+
+// Ping checks to see if the server and measure the time it takes to
+// get back the response.
+func (c *Connection) Ping() (time.Duration, error) {
```

Comment:
I think all of the higher level api calls (Ping, SpaceInfo, SpaceGet, etc) should take a context.Context.

Reply:
done